core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1261,6 +1267,44 @@ class KafkaController(val config: KafkaConfig,
     controlledShutdownCallback(controlledShutdownResult)
   }
 
+  private val topicConfigCache = new BoundedTimedCache[String, Properties](
+    config.controlledShutdownTopicConfigCacheSize,
+    Duration.ofMillis(config.controlledShutdownTopicConfigCacheTTLMs))
+
+  private def fetchTopicConfig(topicName: String): Properties = {
+    val cachedConfig = topicConfigCache.get(topicName)
+    if (cachedConfig != null) {
+      return cachedConfig
+    }
+
+    val config = zkClient.getEntityConfigs(ConfigType.Topic, topicName)
+    topicConfigCache.put(topicName, config)
+    return config
+  }
+
+  private def safeToShutdown(id: Int, brokerEpoch: Long): Boolean = {
+    // If a topic doesn't have min.insync.replicas configured, default to 1
+    val defaultMinISRPropertyValue = "1"
+
+    val atRiskPartitions = controllerContext.partitionsOnBroker(id).filter { partition =>
+      // Look up minISR for this topic, or use the default if not configured.
+      val minISR: Int = fetchTopicConfig(partition.topic())
+        .getOrDefault(TopicConfig.MIN_IN_SYNC_REPLICAS_CONFIG, defaultMinISRPropertyValue)
+        .toString.toInt
+
+      // See which replicas are known alive and not pending shutdown for this partition
+      val liveReplicas = controllerContext.partitionReplicaAssignment(partition).count( { replicaBrokerId =>
```

Comment:
An alive broker could still be out of the ISR. I guess we need to check the latest ISR state.
Also since the info in partitionLeadershipInfo could go stale for 10 - 15 seconds, it's safer to query the partitions' ISR state here.

Reply:
Ah you're right this isn't correct. Thanks.

Reply:
I need to focus on OKR planning but I'll fix this later today.

Reply:
I think this is fixed now.